linked_list/linked_list.py

```diff
@@ -1,79 +1,158 @@
+        while current:
+            if not current.next:
+                return current.value
+            current = current.next   
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: ? O(1)
+    # Space Complexity: ? O(1)
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.head = Node(value)
+            return
+
+        current = self.head
+        # get to the last value in the list
+        while current:
+            if not current.next: #if we're at the last node of the list thennnn
+                # add a new node to the end 
+                current.next = Node(value)
+                return 
+            current = current.next
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: ? O(n)
+    # Space Complexity: ? O(1)
     def find_max(self):
-        pass
+        # check is the list is empty
+        if not self.head:
+            return None
+        # create a varible max_value 
+        max_value = 0
+        # and variable current starting at the head
+        current = self.head
+        # iterate current through from head to last node (for the last node current.next will be falsy None)
+        while current:
+        # compare each current node's value with max_value and update max value if current.value is greater than it
+            if current.value > max_value:
+                max_value = current.value
+            current = current.next
+        # return the max value
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: ? O(n)
+    # Space Complexity: ? O(1)
     def delete(self, value):
         pass
+        # check that the list isn't empty, if it is return none
+        if not self.head:
+            return None
+
+        #if the head matched make .next the head
+        if self.head.value == value:
+            self.head = self.head.next
+            return
+        # iterate thru the list until a node's value == the given value
+        previous = self.head
+        current = self.head.next
+        while current:
+            if current.value == value:
+                previous.next = current.next
+                return
+            previous = current
+            current = current.next
+        return
+        # point the previous node to skip over the matching node and point to the matching mode's .next value
+        # if the matching value is the head, reassign self.head to the next value. 
+        # if there is no next value return None
+        # if theres so match return None
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit(self):

```

Comment:
🪐 Space complexity will actually be O(n) here since you are creating a `helper_list` that will eventually hold every node in the linked list.